### ISSUE TITLE
pgx.Batch behavior when one statement raises an exception

### ISSUE THREAD

**Issue by jschaf** [jackc/pgx#1066]
Hello again, I'd like to clarify the expected behavior of a sequence of pgx batch calls where one of the statements throws an exception in two cases: 1) with explicit transaction commands, 2) with implicit transaction commands.

First case with explicit transactions:
```go
		batch.Queue(`BEGIN`)
		batch.Queue(`INSERT INTO foo (id) VALUES (1)`)
		batch.Queue(`COMMIT`)
		batch.Queue(`SELECT 1/0`) // exception
		batch.Queue(`INSERT INTO foo (id) VALUES (2)`)
```

Second case with implicit transactions:
```go
		batch.Queue(`INSERT INTO foo (id) VALUES (1)`)
		batch.Queue(`SELECT 1/0`) // exception
		batch.Queue(`INSERT INTO foo (id) VALUES (2)`)
```

I wrote up tests with the above commands and found that:

- Execution continues until the first exception, the `SELECT 1/0`. The last insert statement does not run in either case.

- In the first case, the first insert with value `1` inserts successfully since there's an explicit `COMMIT`.
 
- In the second case with implicit transactions, the first insert is rolled back, so nothing is written into the `foo` table.

The Postgres docs for error handling for the extended query protocol states:

> When an error is detected while processing any extended-query message, the backend issues ErrorResponse, then reads and discards messages until a Sync is reached, then issues ReadyForQuery and returns to normal message processing. (But note that no skipping occurs if an error is detected while processing Sync — this ensures that there is one and only one ReadyForQuery sent for each Sync.

I'm not sure how a Sync relates to pgx. My guess is pgx only issues a Sync after sending all queries in a batch.

To summarize:

1. Is a batch wrapped with an implicit transaction block if there's no explicit transaction commands like `BEGIN`?
2. If a statement in a batch fails, will pgx execute any more messages in that batch?

**Comment by jackc** [comment]
> My guess is pgx only issues a Sync after sending all queries in a batch.

Correct.

> Is a batch wrapped with an implicit transaction block if there's no explicit transaction commands like BEGIN?

Correct.

> If a statement in a batch fails, will pgx execute any more messages in that batch?

pgx sends them all. But PostgreSQL will ignore them until it gets the `Sync` message.

**Comment by vorishirne** [comment]
>Is a batch wrapped with an implicit transaction block if there's no explicit transaction commands like BEGIN?

What if I wrap with a BEGIN and END (not BEGIN and COMMIT) statements. In that case too, a batch is wrapped with an implicit transaction?

**Comment by vorishirne** [comment]
I have a procedure defined like this
```sql
CREATE OR REPLACE PROCEDURE update_flow_data_(int,varchar(256)) 
LANGUAGE plpgsql AS $$
	BEGIN 

	UPDATE workers SET email = $2 WHERE id=$1 ;
  	IF NOT FOUND THEN 
    	INSERT INTO workers (id,email) VALUES ($1,$2) ;
  	END IF;
	END;
$$;
```

And I call this procedure in batch like this

`CALL update_op(2,'wrt2');`

Will this be wrapped in transaction or not?

**Comment by jackc** [comment]
A batch that includes explicit begin and commit overrides some of the implicit transactional behavior.

```go
package main

import (
	"context"
	"fmt"
	"log"
	"os"

	"github.com/jackc/pgx/v5"
)

func main() {
	ctx := context.Background()

	conn, err := pgx.Connect(ctx, os.Getenv("DATABASE_URL"))
	if err != nil {
		log.Fatal(err)
	}
	defer conn.Close(ctx)

	_, err = conn.Exec(ctx, "create temporary table t (name text primary key)")
	if err != nil {
		log.Fatal(err)
	}

	batch := &pgx.Batch{}

	batch.Queue("begin")
	batch.Queue("insert into t values ($1)", "foo")
	batch.Queue("commit")
	batch.Queue("select 1 / n from generate_series(-5, 5) n") // cause division by zero error

	err = conn.SendBatch(ctx, batch).Close()
	fmt.Println("batch result:", err)

	var n int32
	err = conn.QueryRow(ctx, "select count(*) from t").Scan(&n)
	if err != nil {
		log.Fatal(err)
	}

	if n == 0 {
		fmt.Println("entire batch was rolled back")
	} else {
		fmt.Println("inner commit stayed committed")
	}
}
```

Results:

```
batch result: ERROR: division by zero (SQLSTATE 22012)
inner commit stayed committed
```

---

As you do not have any explicit transaction control statements in your procedure it should be part of the implicit atomicity of the batch (assuming you don't have other explicit control statements in your batch).

 However, in that particular example I would suggest using `INSERT ... ON CONFLICT` instead of a procedure.

**Comment by vorishirne** [comment]
I was reading this article and it says ON CONFLICT is not fast in case you are expecting 90% updates.
https://www.percona.com/blog/why-insert-on-duplicate-key-update-may-be-slow-by-incurring-disk-seeks/

**Comment by jackc** [comment]
That article is about MySQL, not PostgreSQL. However, it wouldn't greatly surprise me if `INSERT ... ON CONFLICT` was slower than an `UPDATE`. At the very least if your table has a serial / sequence, it is getting advanced.

But the real advantage of `INSERT ... ON CONFLICT` is it's guaranteed correctness even when called concurrently. If your procedure is executed concurrently it is possible both executions would hit the `INSERT` branch. To be correct you would need to have a unique index and have your application code able to handle a uniqueness violation.

Also, if you still want to avoid the `INSERT ... ON CONFLICT`, the entire `UPDATE` or `INSERT` could be encoded into a writable CTE. It still have the concurrency issues of the two statements in a procedure, but it's probably faster and arguably simpler.

**Comment by vorishirne** [comment]
Hi @jackc I have implemented your advice. Still working on the cte part as I am having hard time convince my team why CTEs would be faster. Thanks much.

**Comment by jackc** [comment]
> Hi @jackc I have implemented your advice. Still working on the cte part as I am having hard time convince my team why CTEs would be faster. Thanks much.

I would suggest benchmarking all the different implementations. Go's test package makes that fairly simple. That would provide definite evidence one way or another.

**Comment by vorishirne** [comment]
Yes gonna do that. Will paste the results somewhere.

**Comment by jackc** [comment]
@deepakbhatia

The first statement would be successful, but when the second statement raises an exception the implicit transaction would be rolled back. The third statement would be ignored.

Practically speaking, batches should be used as all or nothing. You can use explicit control statements (e.g. `BEGIN`, `COMMIT`, `ROLLBACK`) in a batch, but they are of limited use. The problem is all statements after an error are ignored. 

See this example:

```go
package main

import (
	"context"
	"fmt"
	"log"
	"os"

	"github.com/jackc/pgx/v5"
)

func main() {
	ctx := context.Background()

	conn, err := pgx.Connect(ctx, os.Getenv("DATABASE_URL"))
	if err != nil {
		log.Fatal(err)
	}
	defer conn.Close(ctx)

	_, err = conn.Exec(ctx, "create temporary table t (name text primary key)")
	if err != nil {
		log.Fatal(err)
	}

	batch := &pgx.Batch{}

	batch.Queue("begin")
	batch.Queue("insert into t values ($1)", "foo")
	batch.Queue("commit")
	batch.Queue("begin")
	batch.Queue("select 1 / n from generate_series(-5, 5) n") // cause division by zero error
	batch.Queue("commit")
	batch.Queue("begin")
	batch.Queue("insert into t values ($1)", "bar")
	batch.Queue("commit")

	err = conn.SendBatch(ctx, batch).Close()
	fmt.Println("batch result:", err)

	// The transaction wrapping the division by zero is still in progress so we need to manually rollback.
	_, err = conn.Exec(ctx, "rollback")
	if err != nil {
		log.Fatal(err)
	}

	var n int32
	err = conn.QueryRow(ctx, "select count(*) from t").Scan(&n)
	if err != nil {
		log.Fatal(err)
	}

	if n == 0 {
		fmt.Println("entire batch was rolled back")
	} else if n == 1 {
		fmt.Println("statements were committed before the error but ignored after the error")
	} else if n == 2 {
		fmt.Println("before and after the error were executed")
	}
}
```

Output:

```
batch result: ERROR: division by zero (SQLSTATE 22012)
statements were committed before the error but ignored after the error
```

You could get more control by going down to the pipeline mode. That gives you control over over implicit transactions by controlling when `Sync` messages are sent. However, pipeline mode is more difficult to use and exposes you to the subtleties of the actual PostgreSQL wire protocol. I wouldn't recommend it unless you *really* need it.
